File: examples/cluster.rs
use cdp_protocol::{cluster::{Cluster, ClusterConfig}, Config, Result};

#[path = "common/logging.rs"]
mod logging;

const URLS: &[&str] = &[
    "https://slishee.com",
    "https://www.rust-lang.org",
    "https://www.google.com",
    "https://github.com",
    "https://stackoverflow.com",
    "https://news.ycombinator.com",
    "https://www.wikipedia.org",
    "https://docs.rs",
    "https://crates.io",
    "https://www.mozilla.org",
];

#[tokio::main]
async fn main() -> Result<()> {
    let _guard = logging::init();

    let cfg = Config::default();
    std::fs::create_dir_all(&cfg.screenshots_dir).ok();

    let config = ClusterConfig {
        concurrency: 3,
        retries: 1,
        monitor: true,
        ..ClusterConfig::from(cfg.clone())
    };

    println!("Starting cluster ({} workers)...", config.concurrency);
    let cluster = Cluster::new(config).await?;

    let shots_dir = cfg.screenshots_dir.clone();
    let results = cluster.run(URLS.iter().copied().enumerate(), move |client, (i, url)| {
        let shots_dir = shots_dir.clone();
        async move {
            client.navigate_and_wait(url, 15_000).await?;
            let title = client.eval("document.title").await?;
            client.full_page_screenshot_to_file(
                &format!("{shots_dir}/cluster_{i:03}.png")
            ).await?;
            Ok::<_, cdp_protocol::CdpError>(title)
        }
    }).await;

    let success = results.iter().filter(|r| r.is_ok()).count();
    let failed  = results.len() - success;

    for (i, r) in results.iter().enumerate() {
        match &r.result {
            Ok(title) => println!("[{i:2}] {title} ({:.1}s)", r.elapsed.as_secs_f64()),
            Err(e)    => println!("[{i:2}] error: {e}"),
        }
    }

    println!("\nDone. {} ok, {} failed.", success, failed);

    cluster.close().await;
    Ok(())
}

File: bindings/node/examples/cluster.mjs
// Cluster: fixed-size worker pool with retries.
// Mirrors examples/cluster.rs.
//
//   google-chrome --headless=new --remote-debugging-port=9222 --no-first-run
//   node examples/cluster.mjs
import { mkdir } from 'node:fs/promises'
import { Cluster } from '../index.js'

const SHOTS = 'screenshots'
await mkdir(SHOTS, { recursive: true })

const URLS = [
  'https://slishee.com',
  'https://www.rust-lang.org',
  'https://www.google.com',
  'https://github.com',
  'https://stackoverflow.com',
  'https://news.ycombinator.com',
  'https://www.wikipedia.org',
  'https://docs.rs',
  'https://crates.io',
  'https://www.mozilla.org',
]

const CONCURRENCY = 3

console.log(`Starting cluster (${CONCURRENCY} workers)...`)
const cluster = await Cluster.create({
  host: '127.0.0.1',
  port: 9222,
  concurrency: CONCURRENCY,
  retries: 1,
})

const tasks = await Promise.all(
  URLS.map((url, i) =>
    cluster.execute([
      { action: 'navigate', url },
      { action: 'wait_for_selector', selector: 'body', timeout_ms: 15_000 },
      { action: 'screenshot', path: `${SHOTS}/cluster_${String(i).padStart(3, '0')}.png` },
      { action: 'get_title' },
    ]),
  ),
)

let ok = 0
tasks.forEach((t, i) => {
  if (t.success) {
    ok++
    console.log(`[${String(i).padStart(2)}] ${t.results.at(-1)?.value} (${(t.elapsedMs / 1000).toFixed(1)}s)`)
  } else {
    console.log(`[${String(i).padStart(2)}] error: ${t.error}`)
  }
})

console.log(`\nDone. ${ok} ok, ${URLS.length - ok} failed.`)
await cluster.close()
